gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java

```diff
@@ -0,0 +1,149 @@
+import java.io.Reader;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is responsible for serializing/deserializing protobuf {@link Message} to/from a proper
+ * format suitable for transferring it over by in a HTTP REST request/response. The information from
+ * a {@code Message} is split between a request body (JSON), URL path parameters and query
+ * parameters.
+ */
+@BetaApi
+public class ProtoRestSerializer<RequestT extends Message> {
+  private ProtoRestSerializer() {}
+
+  /** Creates a new instance of ProtoRestSerializer. */
+  public static <RequestT extends Message> ProtoRestSerializer<RequestT> create() {
+    return new ProtoRestSerializer<>();
+  }
+
+  /**
+   * Serializes the data from {@code message} to a JSON string. The implementation relies on
+   * protobuf native JSON formatter.
+   *
+   * @param message a message to serialize
+   */
+  String toJson(RequestT message) {
+    try {
+      return JsonFormat.printer().print(message);
+    } catch (InvalidProtocolBufferException e) {
+      throw new RuntimeException("Failed to serialize message to JSON", e);
+    }
+  }
+
+  /**
+   * Deserializes a {@code message} from an input stream to a protobuf message.
+   *
+   * @param message the input stream with a JSON-encoded message in it
+   * @param builder an empty builder for the specific {@code RequestT} message to serialize
+   */
+  @SuppressWarnings("unchecked")
+  RequestT fromJson(InputStream message, Message.Builder builder) {
+    try (Reader json = new InputStreamReader(message)) {
+      JsonFormat.parser().ignoringUnknownFields().merge(json, builder);
+      return (RequestT) builder.build();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to parse response message", e);
+    }
+  }
+
+  /**
+   * Puts a message field in {@code fields} map which will be used to populate URL path of a
+   * request.
+   *
+   * @param fields a map with serialized fields
+   * @param fieldName a field name
+   * @param fieldValue a field value
+   */
+  public void putPathParam(Map<String, String> fields, String fieldName, Object fieldValue) {
+    if (isDefaultValue(fieldName, fieldValue)) {
+      return;
+    }
+    fields.put(fieldName, String.valueOf(fieldValue));
+  }
+
+  /**
+   * Puts a message field in {@code fields} map which will be used to populate query parameters of a
+   * request.
+   *
+   * @param fields a map with serialized fields
+   * @param fieldName a field name
+   * @param fieldValue a field value
+   */
+  public void putQueryParam(Map<String, List<String>> fields, String fieldName, Object fieldValue) {
+    // Avoids empty query parameter
+    if (isDefaultValue(fieldName, fieldValue)) {
+      return;
+    }
+
+    ImmutableList.Builder<String> paramValueList = ImmutableList.builder();
+    if (fieldValue instanceof List<?>) {
+      for (Object fieldValueItem : (List<?>) fieldValue) {
+        paramValueList.add(String.valueOf(fieldValueItem));
```

Comment:
Optional: Curious whether gax-java is still on Java 7, otherwise we could use a forEach here.

Reply:
GAX is still on Java 7, as are all GCP libraries. See go/nojava7